### PR TITLE
Yeet the `effects` feature, move it onto `const_trait_impl`

### PR DESCRIPTION
**Pull request by compiler-errors**
This PR merges the `effects` feature into the `const_trait_impl` feature. There's really no need to have two feature gates for one feature.

After this PR, if `const_trait_impl` **is** enabled:
* Users can use and define const traits
* `HostEffect` const conditions will be enforced on the HIR
* We re-check the predicates in MIR just to make sure that we don't "leak" anything during MIR lowering

And if `const_trait_impl` **is not** enabled:
* Users cannot use nor define const traits
* `HostEffect` const conditions are not enforced on the HIR
* We will raise a const validation error if we call a function that has any const conditions (i.e. const traits and functions with any `~const` in their where clasues)

This should be the last step for us to be able to enable const traits in the standard library. We still need to re-constify `Drop` and `Destruct` and stuff for const traits to be particularly *useful* for some cases, but this is a good step :D

r? fee1-dead
cc @rust-lang/project-const-traits

### CODE REVIEW

**Comment by compiler-errors** [comment]
Lots of files changed, but most of it is just removing `feature(effects)` from the ui tests, and renaming `FIXME(effects)` to `FIXME(const_trait_impl)` ✨

File: tests/rustdoc/rfc-2632-const-trait-impl.rs [hunk 1]
```diff
@@ -6,10 +6,10 @@
 // stabilized when changing `@!has` to `@has`, and please do
 // not remove this test.
 //
-// FIXME(effects) add `const_trait` to `Fn` so we use `~const`
-// FIXME(effects) restore `const_trait` to `Destruct`
+// FIXME(const_trait_impl) add `const_trait` to `Fn` so we use `~const`
+// FIXME(const_trait_impl) restore `const_trait` to `Destruct`
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/rustdoc/rfc-2632-const-trait-impl.rs:11]:
```suggestion
```

File: tests/ui/consts/auxiliary/closure-in-foreign-crate.rs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 //@ compile-flags: -Znext-solver
 #![crate_type = "lib"]
-#![feature(const_closures, const_trait_impl, effects)]
+#![feature(const_closures, const_trait_impl)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/consts/auxiliary/closure-in-foreign-crate.rs:4]:
```suggestion
```

File: tests/ui/consts/const-try.rs [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 #![crate_type = "lib"]
 #![feature(try_trait_v2)]
-#![feature(const_trait_impl, effects)]
+#![feature(const_trait_impl)]
 #![feature(const_try)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/consts/const-try.rs:9]:
```suggestion
```

File: tests/ui/consts/const_cmp_type_id.rs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 //@ compile-flags: -Znext-solver
-#![feature(const_type_id, const_trait_impl, effects)]
+#![feature(const_type_id, const_trait_impl)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/consts/const_cmp_type_id.rs:3]:
```suggestion
```

File: tests/ui/consts/rustc-impl-const-stability.rs [hunk 5]
```diff
@@ -2,7 +2,7 @@
 //@ known-bug: #110395
 
 #![crate_type = "lib"]
-#![feature(staged_api, const_trait_impl, effects)]
+#![feature(staged_api, const_trait_impl)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/consts/rustc-impl-const-stability.rs:6]:
```suggestion
```

File: tests/ui/delegation/unsupported.rs [hunk 6]
```diff
@@ -1,6 +1,5 @@
 #![feature(const_trait_impl)]
 #![feature(c_variadic)]
-#![feature(effects)]
 #![feature(fn_delegation)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/delegation/unsupported.rs:4]:
```suggestion
```

File: tests/ui/intrinsics/safe-intrinsic-mismatch.rs [hunk 7]
```diff
@@ -1,10 +1,5 @@
-//@ revisions: stock effects
 #![feature(intrinsics)]
 #![feature(rustc_attrs)]
-// as effects insert a const generic param to const intrinsics,
-// check here that it doesn't report a const param mismatch either
-// enabling or disabling effects.
-#![cfg_attr(effects, feature(effects))]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/intrinsics/safe-intrinsic-mismatch.rs:3]:
```suggestion
```

File: tests/ui/traits/const-traits/call.rs [hunk 8]
```diff
@@ -1,6 +1,6 @@
-// FIXME(effects) check-pass
+// FIXME(const_trait_impl) check-pass
 //@ compile-flags: -Znext-solver
-#![feature(const_closures, const_trait_impl, effects)]
+#![feature(const_closures, const_trait_impl)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/call.rs:4]:
```suggestion
```

File: tests/ui/traits/const-traits/cross-crate.rs [hunk 9]
```diff
@@ -1,7 +1,7 @@
 //@ revisions: stock gated stocknc gatednc
 //@ [gated] check-pass
 //@ compile-flags: -Znext-solver
-#![cfg_attr(any(gated, gatednc), feature(const_trait_impl, effects))]
+#![cfg_attr(any(gated, gatednc), feature(const_trait_impl))]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/cross-crate.rs:5]:
```suggestion
```

File: tests/ui/traits/const-traits/do-not-const-check-override.rs [hunk 10]
```diff
@@ -1,7 +1,7 @@
 //@ check-pass
 //@ compile-flags: -Znext-solver
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/do-not-const-check-override.rs:3]:
```suggestion
```

File: tests/ui/traits/const-traits/effects/const_closure-const_trait_impl-ice-113381.rs [hunk 11]
```diff
@@ -1,5 +1,5 @@
 //@ compile-flags: -Znext-solver
-#![feature(const_closures, const_trait_impl, effects)]
+#![feature(const_closures, const_trait_impl)]
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/effects/const_closure-const_trait_impl-ice-113381.rs:3]:
```suggestion
```

File: tests/ui/traits/const-traits/ice-119717-constant-lifetime.rs [hunk 12]
```diff
@@ -1,5 +1,5 @@
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/ice-119717-constant-lifetime.rs:1]:
```suggestion
```

File: tests/ui/traits/const-traits/ice-126148-failed-to-normalize.rs [hunk 13]
```diff
@@ -1,5 +1,5 @@
 #![allow(incomplete_features)]
```
Comment (fee1-dead) [inline tests/ui/traits/const-traits/ice-126148-failed-to-normalize.rs:1]:
```suggestion
```

**Review by fee1-dead — CHANGES_REQUESTED** [review]
r=me after nits and rebases that become necessary

**Comment by compiler-errors** [comment]
The `const_closure` and some other feature gates remain incomplete, so there are still a *few* effects tests that still need `allow(incomplete_features)`. I tried to remove the rest though.

@bors r=fee1-dead

**Comment by RalfJung** [comment]
> This should be the last step for us to be able to enable const traits in the standard library. 

Do we have checks ensuring that const trait fns cannot be called from stable const fns? I don't think so (just some leftovers in check.rs you said you were not happy with).

IMO that is a blocker for enabling the feature in the standard library.

File: compiler/rustc_const_eval/src/check_consts/check.rs [hunk 14]
```diff
@@ -663,8 +657,9 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                     // typeck ensures the conditions for calling a const trait method are met,
                     // so we only error if the trait isn't const. We try to resolve the trait
                     // into the concrete method, and uses that for const stability checks.
-                    // FIXME(effects) we might consider moving const stability checks to typeck as well.
-                    if tcx.features().effects() && trait_is_const {
+                    // FIXME(const_trait_impl) we might consider moving const stability checks
+                    // to typeck as well.
+                    if tcx.features().const_trait_impl() && trait_is_const {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/check_consts/check.rs:662]: The `effects` feature name still appears a few lines down in this file

https://github.com/rust-lang/rust/blob/6b96103bf3b0746e443f9bb86b532be2c536113a/compiler/rustc_const_eval/src/check_consts/check.rs#L680-L684

**Comment by RalfJung** [comment]
> Do we have checks ensuring that const trait fns cannot be called from stable const fns? I don't think so (just some leftovers in check.rs you said you were not happy with, and I don't know if we can trust them).

So specifically, I am concerned about this line

https://github.com/rust-lang/rust/blob/6b96103bf3b0746e443f9bb86b532be2c536113a/compiler/rustc_const_eval/src/check_consts/check.rs#L662

This does not take into account whether the current function must be recursively const-stable.

So I think what the logic here should do is more like
```rust
if let Some(trait_did) = tcx.trait_of_item(callee) {
  if !trait_is_const {
    // Trigger `ops::FnCallNonConst`.
    // ...
    return;
  }

  // Ensure we have `const_trait_impl` and are allowed to use it in this function.
  self.check_op(ops::ConstTraitCall);

  // We don't know which function is being called here, so we can't run the checks below.
  // FIXME: apply more fine-grained per-trait const stability checks
  // (see https://github.com/rust-lang/project-const-traits/issues/5).
  return;
}
```

Also it seems a bit odd that `const_trait_impl` guards *calling* const trait functions. Maybe `const_trait` would make more sense as the name for this feature?

**Comment by fee1-dead** [comment]
> Do we have checks ensuring that const trait fns cannot be called from stable const fns? I don't think so (just some leftovers in check.rs you said you were not happy with, and I don't know if we can trust them).

noting here for completeness, that was done in #132823.

**Pull request merged**